Local relay implementation

rafasofizada opened the pull request:
Signed-off-by: rafasofizada <rafa.sofizadeh@gmail.com>

Hexxeh commented on packages/local-relay/src/index.ts:
```diff
@@ -0,0 +1,25 @@
+
+  writeFile(
+    relayPidFilePath,
+    JSON.stringify({ port, pid: process.pid }),
```
Duplicates line 8?

Hexxeh commented on packages/local-relay/src/RelayServer.ts:
```diff
@@ -0,0 +1,220 @@
+
+    for (const role of roles) {
+      if (!role.match(roleRegex)) {
+        // Should I throw?
```
If you catch in the caller, sure.

rafasofizada replied:
`https://github.com/Fitbit/developer-bridge/pull/182/commits/ae7fbbeef021e4cc8bf76d1f3c756c13ee8e45f2`

Hexxeh commented on packages/local-relay/src/RelayServer.ts:
```diff
@@ -0,0 +1,220 @@
+        '/1/user/-/developer-relay/hosts/:id',
+        this.createConnectionURL.bind(this),
+      )
+      .delete('/connections/:connectionId', this.closeConnection.bind(this));
```
I don't think this is necessary in a local relay, it's intended for cross-relay-node connection cleanup.

Hexxeh commented on packages/local-relay/package.json:
```diff
@@ -0,0 +1,50 @@
+    "/lib/**/*.json"
+  ],
+  "engines": {
+    "node": ">=8.6.0"
```
This can safely be >=12.0.0, last LTS.

rafasofizada replied:
`https://github.com/Fitbit/developer-bridge/pull/182/commits/fa15c8c6999dae5376d7440d49487249ae424bc6`

rafasofizada committed:
Minor

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Update yarn.lock

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

Hexxeh commented on packages/local-relay/package.json:
```diff
@@ -0,0 +1,50 @@
+  "author": "Fitbit, Inc.",
+  "homepage": "https://github.com/Fitbit/developer-bridge/tree/main/packages/local-relay#readme",
+  "license": "BSD-3-Clause",
+  "main": "lib/local-relay.js",
```
How does this file get built? I don't see a matching file in sources.

rafasofizada replied:
`https://github.com/Fitbit/developer-bridge/pull/182/commits/fa15c8c6999dae5376d7440d49487249ae424bc6`

Hexxeh commented on packages/local-relay/package.json:
```diff
@@ -0,0 +1,50 @@
+    "url": "https://github.com/Fitbit/developer-bridge/issues"
+  },
+  "devDependencies": {
+    "@babel/core": "^7.16.0",
```
What requires Babel?

Hexxeh commented on packages/local-relay/src/Host.ts:
```diff
@@ -0,0 +1,84 @@
+  }
+
+  canConnectPeer() {
+    // ASK: Error throw, or return false? I'd consider an error throw
```
As long as you handle it correctly and send an appropriate respones, sure.

rafasofizada replied:
`https://github.com/Fitbit/developer-bridge/pull/182/commits/ae7fbbeef021e4cc8bf76d1f3c756c13ee8e45f2`

Hexxeh commented on packages/local-relay/src/RelayServer.test.ts:
```diff
@@ -0,0 +1,228 @@
+import supertest from 'supertest';
+import * as websocket from 'websocket';
+import Host from './Host';
```
nit: group relative and non-relative imports

rafasofizada committed:
Improved relayInfo, launch test coverage

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
RelayServer handle getRoles error

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
HostStore tests

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Connection tests

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Minor RelayServer tests

Lots of lines shown as "uncovered" in the coverage report actually are; don't know how to make it noticeable for jest

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
canConnectPeer fn signature

canConnectPeer should answer the question straightforward – is it possible to connect a peer or not? Not having an existing connection means "false", and doesn't warrant an exception. Exceptions make sense in connectPeer() – there throwing an exception on non-existent connection makes sense.

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

Hexxeh commented:
I see there's some CI issues and unresolved comments, ping me via a comment once this is ready for review. Also I sent you some info via hangouts. :)

rafasofizada committed:
local-relay package.json update

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
TSLint style fix

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Remove @babel/core from devDep

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Work around nock error

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
yarn.lock remove babel

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

Hexxeh commented on packages/local-relay/src/RelayServer.ts:
```diff
@@ -0,0 +1,210 @@
+
+      const host: Host = this.hostStore.addOrReplace({
+        id: hostId,
+        // tslint:disable-next-line:object-shorthand-properties-first
```
Why not just do what the linter is asking rather than disable it here?

rafasofizada replied:
I strongly prefer putting "primary keys" at the top of "queries"

Hexxeh replied:
The linter is in place to keep everyone's code consistent, so please follow it even if your preferences differ.

rafasofizada replied:
Oh I'm sorry, I always forget that TSLint rules are __explicitly set by someone__ and am used to overriding them when needed :)  Changed

rafasofizada committed:
Code style

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>
